mlkem/rej_uniform.c

```diff
@@ -0,0 +1,369 @@
+    {0, 1, 2, 3, 6, 7, 10, 11, 12, 13, 14, 15, -1, -1, -1, -1},       // 235
+    {4, 5, 6, 7, 10, 11, 12, 13, 14, 15, -1, -1, -1, -1, -1, -1},     // 236
+    {0, 1, 4, 5, 6, 7, 10, 11, 12, 13, 14, 15, -1, -1, -1, -1},       // 237
+    {2, 3, 4, 5, 6, 7, 10, 11, 12, 13, 14, 15, -1, -1, -1, -1},       // 238
+    {0, 1, 2, 3, 4, 5, 6, 7, 10, 11, 12, 13, 14, 15, -1, -1},         // 239
+    {8, 9, 10, 11, 12, 13, 14, 15, -1, -1, -1, -1, -1, -1, -1, -1},   // 240
+    {0, 1, 8, 9, 10, 11, 12, 13, 14, 15, -1, -1, -1, -1, -1, -1},     // 241
+    {2, 3, 8, 9, 10, 11, 12, 13, 14, 15, -1, -1, -1, -1, -1, -1},     // 242
+    {0, 1, 2, 3, 8, 9, 10, 11, 12, 13, 14, 15, -1, -1, -1, -1},       // 243
+    {4, 5, 8, 9, 10, 11, 12, 13, 14, 15, -1, -1, -1, -1, -1, -1},     // 244
+    {0, 1, 4, 5, 8, 9, 10, 11, 12, 13, 14, 15, -1, -1, -1, -1},       // 245
+    {2, 3, 4, 5, 8, 9, 10, 11, 12, 13, 14, 15, -1, -1, -1, -1},       // 246
+    {0, 1, 2, 3, 4, 5, 8, 9, 10, 11, 12, 13, 14, 15, -1, -1},         // 247
+    {6, 7, 8, 9, 10, 11, 12, 13, 14, 15, -1, -1, -1, -1, -1, -1},     // 248
+    {0, 1, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, -1, -1, -1, -1},       // 249
+    {2, 3, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, -1, -1, -1, -1},       // 250
+    {0, 1, 2, 3, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, -1, -1},         // 251
+    {4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, -1, -1, -1, -1},       // 252
+    {0, 1, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, -1, -1},         // 253
+    {2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, -1, -1},         // 254
+    {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15},           // 255
+};
+
+const uint16_t bit_table[8] = {0x1, 0x2, 0x4, 0x8, 0x10, 0x20, 0x40, 0x80};
+/*************************************************
+* Name:        rej_uniform_large
+*
+* Description: Rejection sampling on uniform random bytes to generate
+*              uniform random integers mod q
+*
+* Arguments:   - int16_t *r:          pointer to output buffer
+*              - unsigned int len:    requested number of 16-bit integers
+*                                     (uniform mod q)
+*              - const uint8_t *buf:  pointer to input buffer
+*                                     (assumed to be uniform random bytes)
+*              - unsigned int buflen: length of input buffer in bytes
+*              - unsigned int *consumed: the length of consumed buffer
+* Returns number of sampled 16-bit integers (at most len)
+**************************************************/
+static unsigned int rej_uniform_large(int16_t *r, unsigned int len,
+                                     const uint8_t *buf, unsigned int buflen,
+                                     unsigned int *consumed) {
+
+    // Ensure alignment, the minimum stack size is (len + 32) * sizeof(int16_t)
+    // Since we don't want to use malloc/free, we assume the maximum len = KYBER_N and use stack.
+    int16_t stack[KYBER_N + 32] = {0};
```

Comment:
There's no need to do this here -- you can allocate stack as needed in the ASM?

Reply:
You're right, I'm thinking of using `memcpy` to copy data from stack to `r`,  so I forgot to move the stack and memcpy to asm.

Reply:
done in https://github.com/pq-code-package/mlkem-c-aarch64/pull/47/commits/3cf987be5dc19702a271ee94a7585fd202fde681